### PR TITLE
OSD-11792 Migrate to OperatorSDK v1.21

### PR DESCRIPTION
**Pull request by mrbarge**
### What type of PR is this?
Feature

### What this PR does / why we need it?

This PR migrates the `splunk-forwarder-operator` to Operator SDK v1.21 and bumps/aligns its other dependencies (ie. `controller-runtime`) as part of SDE-1972.

No functional changes have been made to the operator's underlying duties. Some changes have been done to the controllers and `main.go` which are required to align with SDK v1.x conventions.

A more specific summary of changes:
- Dependencies:
  - OperatorSDK: Moved to version 1.21 (from 0.16.0)
  - controller-runtime: Moved to 0.12.3 (from 0.4.0)
- Boilerplate: bumped to latest version
- Changes:
  - OperatorSDK no longer has built-in metrics Service creation functions, so this was done by alternate means in `main.go` instead.
    - `operator-custom-metrics` was not used for this, because the operator did not contain any custom local metrics to publish.
    - the existing operator does not create a ServiceMonitor, so this behaviour was preserved, although it can be toggled on if you'd like.
  - API is now available at `github.com/openshift/splunk-forwarder-operator/api/v1alpha1` per SDK conventions
  - Controller reconciler structs renamed to `SplunkForwarderReconciler` and `SecretReconciler` per SDK conventions
  - Secret Controller predicate changed to a function (`mySecretPredicate`) per SDK conventions of how predicates are used.
  - `config/config.go`
    - `OperatorNamespace` was updated to the correct value (note: this variable was not being used anywhere in the existing operator, but it is now used for the Service creation)
  - `version/version.go`
    - Added `SDKVersion` variable to define the last-verified SDK version used (this information is not available in SDK v1.x)
  - `secret_controller_test.go` / `splunkforwarder_controller_test.go`
    - Updated to build the fake client using `NewClientBuilder` rather than the deprecated `NewFakeClientWithScheme` as this threw lint deprecation errors.

### Which Jira/Github issue(s) this PR fixes?

[OSD-11792](https://issues.redhat.com//browse/OSD-11792)

### Notes for the reviewer

A thorough series of tests has been undertaken to verify that the migrated operator is behaving correctly. The test plan and tests conducted can be reviewed in the comments of testing ticket OSD-12774.

It will be independently verified by another member of the team.

### CODE REVIEW

**Comment by mrbarge** [comment]
/hold
whilst being reviewed

File: main.go
Comment (mjlshen) [inline main.go:104]: This seems strange, couldn't we control this behavior with the `leader-elect`/`enableLeaderElection` flag/bool when creating the new manager?

File: go.mod
Comment (mjlshen) [inline go.mod:9]: I think we wanted to remain on `k8s.io/* v0.23.5` and `controller-runtime v0.11.2` for now per [SDE-1972](https://issues.redhat.com/browse/SDE-1972)

File: controllers/splunkforwarder/splunkforwarder_controller.go
Comment (mjlshen) [inline controllers/splunkforwarder/splunkforwarder_controller.go:65]: Leftover TODO :D

File: main.go
Comment (mrbarge) [reply main.go:104]: The `leader-elect` flag, when passed to the controller manager, will set it up to perform leader-with-lease election. Whereas using `Become()` will use leader-for-life election.

Either option is validly supported in the SDK, and neither is the "right" answer - SDK 1.x defaults to lease, but leaves the decision ultimately up to the operator author. Currently, SFO performs leader-for-life, so as part of SDE-1972 we've been trying to avoid subtly changing operator behaviour unless necessary - hence maintaining leader-for-life here.

Ultimately, given the operator runs 1 replica, I suspect whichever method is chosen is largely a non-issue. But I'll leave that decision up to yourselves as the operator custodians. :)  Happy to switch to a leader-with-lease approach if you'd prefer that, and remove this block - just let me know if that suits.

File: go.mod
Comment (mrbarge) [reply go.mod:9]: I see that goal as more of a "let's bring everything up to this as the bare minimum", as there will inevitably be drift. (cc @rendhalver - your thoughts?) I can always lock in these versions in the `replace`, though.

File: controllers/splunkforwarder/splunkforwarder_controller.go
Comment (mrbarge) [reply controllers/splunkforwarder/splunkforwarder_controller.go]: Thanks for spotting! Fixed now.

File: main.go
Comment (mrbarge) [reply main.go:104]: I will note that, having just attempted to test the switch to leader-by-lease, I can observe that it would also require the operator serviceaccount's RBAC to require additional permissions, because currently it will fail with:

```
E0815 22:28:12.477163       1 leaderelection.go:330] error retrieving resource lock openshift-splunk-forwarder-operator/f48c54da.managed.openshift.io: leases.coordination.k8s.io "f48c54da.managed.openshift.io" is forbidden: User "system:serviceaccount:openshift-splunk-forwarder-operator:splunk-forwarder-operator" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "openshift-splunk-forwarder-operator"
```

Comment (mjlshen) [reply main.go:104]: Thanks for the information! Honestly I am not the most familiar with SFO haha, though reviewing the documentation I agree with your assessment that this is likely a non-issue either way

File: go.mod
Comment (mjlshen) [reply go.mod:9]: Got it, I'm also in favor of this sentiment (and not a fan of having to `replace` unless we absolutely must haha) - thanks for explaining!

**Review by mjlshen — APPROVED** [review]
/lgtm

Potential discussion remaining on whether to pin this down to K8s 0.23, but the changes look good to me - thanks for this!

**Comment by mrbarge** [comment]
Thanks a lot for the review @mjlshen !

Once one of my teammates has verified/repeated my testing results, I'll remove the hold on the PR.

**Comment by mrbarge** [comment]
A small additional update has been added to this PR to add some documentation to the `README` around build/development dependencies. (refs: OSD-12776)

**Comment by ravitri** [comment]
Based on [OSD-12774](https://issues.redhat.com/browse/OSD-12774), I tested SFO's migrated operator on 4.11 and 4.6 clusters and can confirm that it functionally worked as expected. With that, it's LGTM from my end too. What's pending is e2e test which we can monitor after this PR is merged.

**Pull request merged**
